File: main/n-ary-tree-postorder-traversal/n-ary-tree-postorder-traversal.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
public:
    static vector<int> postorder(const Node* root);
};

vector<int> Solution::postorder(const Node* const root)
{
    vector<int> ret;
    
    stack<const Node*> s;
    s.push(root);
    
    while (!s.empty()) {
        const auto parent = s.top();
        s.pop();
        if (!parent) continue;
        
        ret.push_back(parent->val);
        for (const auto child : parent->children) s.push(child);
    }
    
    reverse(begin(ret), end(ret));
    return ret;
}

File: main/n-ary-tree-postorder-traversal/n-ary-tree-postorder-traversal.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    public List<Integer> postorder(final Node root) {
        final List<Integer> ret = new ArrayList<>();
        if (root == null) return ret;

        final Deque<Node> stack = new ArrayDeque<>();
        stack.push(root);

        while (!stack.isEmpty()) {
            final Node node = stack.pop();
            ret.add(node.val);
            for (final Node child : node.children) stack.push(child);
        }

        Collections.reverse(ret);
        return ret;
    }
}
